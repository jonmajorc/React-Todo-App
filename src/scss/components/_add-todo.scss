.add-todo {
  &__btn {
    margin: 0 5px;

    &--add {
      background-color: $primary;
      margin: 0 5px;
    }
    &--cancel {
      background-color: $red;
      margin: 0 5px;
    }
  }
}
