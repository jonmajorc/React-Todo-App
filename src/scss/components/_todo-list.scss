.todo-list {
  list-style: none;

  &__item {
    padding: 5px 0;
    border-bottom: 1px solid lighten($grey, 35%);

    &__text {
      &[data-completed='true'] {
        text-decoration: line-through;
        color: $red;
      }
    }
    &__help {
      color: $grey;
    }

    &__btn--del {
      border-radius: 50%;
      color: $grey;

      &:hover {
        text-decoration: underline;
      }
    }
  }
}
