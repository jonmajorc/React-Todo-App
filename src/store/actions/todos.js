import actionTypes from '../actionTypes'

export const addTodo = todo => ({
  type: actionTypes.ADD_TODO,
  todo,
})

export const toggleTodo = (id, isComplete) => ({
  type: actionTypes.TOGGLE_TODO,
  isComplete,
  id,
})

export const removeTodo = id => ({
  type: actionTypes.REMOVE_TODO,
  id,
})
