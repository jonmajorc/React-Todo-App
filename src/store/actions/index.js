import * as todoActions from './todos'

export default {
  ...todoActions,
}