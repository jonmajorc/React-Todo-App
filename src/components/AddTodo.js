import React, { Component } from 'react'
import { connect } from 'react-redux'
import actions from '../store/actions'
import PropTypes from 'prop-types'

export class AddTodo extends Component {
  static propTypes = {
    todos: PropTypes.array.isRequired,
    addTodo: PropTypes.func.isRequired,
  }

  state = {
    id: null,
    todo: '',
    isComplete: false,
  }

  handleTodoInput = e => {
    const todo = e.target.value
    this.setState({
      todo,
    })
  }

  handleAddTodo = e => {
    if ((e.key !== 'Enter' && e.type !== 'click') || !this.state.todo) return
    this.props.addTodo({ ...this.state, id: this.nextTodoId })
  }

  render() {
    const { onCancel } = this.props

    return (
      <div data-test="add-todo" className="add-todo">
        <input
          type="text"
          onChange={this.handleTodoInput}
          value={this.state.todo}
          data-test="add-todo-input"
          className="add-todo__input"
          onKeyPress={this.handleAddTodo}
        />
        <button
          data-test="add-todo-btn"
          className="add-todo__btn--add"
          onClick={this.handleAddTodo}
        >
          Add
        </button>
        {onCancel && (
          <button
            data-test="cancel-todo-btn"
            className="add-todo__btn--cancel"
            onClick={onCancel}
          >
            cancel
          </button>
        )}
      </div>
    )
  }
  get nextTodoId() {
    const { todos } = this.props
    const todoLen = todos.length
    const todo = todos[todoLen ? todoLen - 1 : 0]
    if (todo) {
      return todo.id + 1
    }
    return 0
  }
}

const mapStateToProps = ({ todos }) => ({
  todos,
})

const { addTodo } = actions

export default connect(mapStateToProps, {
  addTodo,
})(AddTodo)
