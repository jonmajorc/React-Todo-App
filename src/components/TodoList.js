import React, { Component } from 'react'
import { connect } from 'react-redux'
import actions from '../store/actions'
import PropTypes from 'prop-types'

export class TodoList extends Component {
  static propTypes = {
    todos: PropTypes.array.isRequired,
    toggleTodo: PropTypes.func.isRequired,
    removeTodo: PropTypes.func.isRequired,
  }
  render() {
    const { todos, toggleTodo, removeTodo } = this.props
    return (
      <ul className="todo-list">
        {todos.map(({ id, todo, isComplete }) => (
          <li
            data-test={`todo-item-${id}`}
            className="todo-list__item"
            key={id}
            onClick={() => toggleTodo(id, isComplete)}
          >
            <button
              className="todo-list__item__btn--del"
              onClick={e => {
                e.stopPropagation()
                removeTodo(id)
              }}
            >
              X
            </button>
            <span className="todo-list__item__text" data-completed={isComplete}>
              {todo}
            </span>
          </li>
        ))}
      </ul>
    )
  }
}

const mapStateToProps = ({ todos }) => ({
  todos,
})

const { removeTodo, toggleTodo } = actions

export default connect(mapStateToProps, {
  removeTodo,
  toggleTodo,
})(TodoList)
