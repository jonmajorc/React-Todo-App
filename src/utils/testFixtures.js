// modules
import React from 'react'
import { mount } from 'enzyme'
// import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import { configureStoreTest } from '../store/configureStore'
import initialState from '../store/initialState'
import { createStore } from 'redux'
import rootReducer from '../store/reducers'
import { MemoryRouter as Router } from 'react-router-dom'
// components
import App from '../layouts/App'

/***************************************************************************\
  Object Mothers
\***************************************************************************/
export function renderAppComponent(props = {}, opts) {
  const { route = '', Comp = App } = opts

  const initialStoreState = {
    ...initialState,
    ...props,
  }

  let store, container

  const mockStore = configureStoreTest()
  store = createStore(rootReducer, initialStoreState)
  return {
    store,
    mount: mount(
      <Provider store={store}>
        <Router initialEntries={[route]}>
          <Comp />
        </Router>
      </Provider>
    ),
  }
}
