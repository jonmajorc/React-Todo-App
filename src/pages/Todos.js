// modules
import React, { Component } from 'react'

// components
import AddTodo from '../components/AddTodo'
import TodoList from '../components/TodoList'

export class Todos extends Component {
  state = {
    isAdding: false,
  }

  onAddTodo = () => {
    this.setState(prevState => ({
      isAdding: !prevState.isAdding,
    }))
  }

  onCancel = () => {
    this.setState(prevState => ({
      isAdding: false,
    }))
  }

  render() {
    return (
      <div data-test="todo-page" className="todos">
        <header className="todos__header">
          <h3>My Todos</h3>
          <small className="todos__header__date">
            {new Date().toDateString()}
          </small>
          {this.state.isAdding ? (
            <AddTodo onCancel={this.onCancel} />
          ) : (
            <button
              className="todos__new-todo-btn"
              data-test="new-todo-btn"
              onClick={this.onAddTodo}
            >
              + Add Todo
            </button>
          )}
        </header>
        <small className="todos__header__help">
          (&nbsp;Click a todo to complete it!&nbsp;)
        </small>
        <TodoList />
      </div>
    )
  }
}
export default Todos
