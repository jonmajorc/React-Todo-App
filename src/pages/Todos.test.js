import React from 'react'

// utils
import { renderAppComponent } from '../utils/testFixtures'
import toJson from 'enzyme-to-json'
// components
import { AddTodo } from '../components/AddTodo'
import { TodoList } from '../components/AddTodo'

// pages
import { Todo } from './Todos'
// redux
import actions from '../store/actions'

describe('Todos Page', () => {
  it('renders todo page', () => {
    const todos = []
    const todoPage = renderAppComponent(
      {
        todos,
      },
      { route: '/todos' }
    ).mount
    expect(todoPage.find('[data-test="todo-page"]')).toMatchSnapshot()
  })

  it.only('inputs new todo item in input field with existing items', () => {
    const todos = [
      {
        id: 0,
        todo: 'clean car',
        isComplete: false,
      },
      {
        id: 1,
        todo: 'brush teeth',
        isComplete: false,
      },
    ]
    const addTodo = 'clean my room'

    const container = renderAppComponent(
      {
        todos,
      },
      { route: '/todos' }
    )
    const TodoPage = container.mount
    const newTodoBtn = TodoPage.find('[data-test="new-todo-btn"]')

    newTodoBtn.simulate('click')
    const AddTodoComponent = container.mount.find(AddTodo)
    const addTodoInput = AddTodoComponent.find('[data-test="add-todo-input"]')
    const addTodoBtn = AddTodoComponent.find('[data-test="add-todo-btn"]')

    addTodoInput.simulate('change', { target: { value: addTodo } })

    expect(addTodoInput.instance().value).toEqual(addTodo)

    addTodoBtn.simulate('click')

    expect(container.store.getState().todos).toHaveLength(3)
    expect(container.store.getState().todos[2].id).toEqual(2)
  })

  it('inputs new todo item in input field', () => {
    const todos = []
    const addTodo = 'clean my room'

    const container = renderAppComponent(
      {
        todos,
      },
      { route: '/todos' }
    )

    const TodoPage = container.mount
    const newTodoBtn = TodoPage.find('[data-test="new-todo-btn"]')

    newTodoBtn.simulate('click')
    const AddTodoComponent = container.mount.find(AddTodo)

    const addTodoInput = AddTodoComponent.find('[data-test="add-todo-input"]')
    const addTodoBtn = AddTodoComponent.find('[data-test="add-todo-btn"]')

    addTodoInput.simulate('change', { target: { value: addTodo } })

    expect(addTodoInput.instance().value).toEqual(addTodo)

    addTodoBtn.simulate('click')

    expect(container.store.getState().todos).toHaveLength(1)
    expect(container.store.getState().todos[0].id).toEqual(0)
  })

  it('inputs new todo item in input field with existing items', () => {
    const todos = [
      {
        id: 0,
        todo: 'clean car',
        isComplete: false,
      },
      {
        id: 1,
        todo: 'brush teeth',
        isComplete: false,
      },
    ]
    const addTodo = 'clean my room'

    const container = renderAppComponent(
      {
        todos,
      },
      { route: '/todos' }
    )
    const TodoPage = container.mount
    const newTodoBtn = TodoPage.find('[data-test="new-todo-btn"]')

    newTodoBtn.simulate('click')
    const AddTodoComponent = container.mount.find(AddTodo)
    const addTodoInput = AddTodoComponent.find('[data-test="add-todo-input"]')
    const addTodoBtn = AddTodoComponent.find('[data-test="add-todo-btn"]')

    addTodoInput.simulate('change', { target: { value: addTodo } })

    expect(addTodoInput.instance().value).toEqual(addTodo)

    addTodoBtn.simulate('click')

    expect(container.store.getState().todos).toHaveLength(3)
    expect(container.store.getState().todos[2].id).toEqual(2)
  })

  it('inputs new todo item in input field', () => {
    const todos = []
    const addTodo = 'clean my room'

    const container = renderAppComponent(
      {
        todos,
      },
      { route: '/todos' }
    )

    const TodoPage = container.mount
    const newTodoBtn = TodoPage.find('[data-test="new-todo-btn"]')

    newTodoBtn.simulate('click')
    const AddTodoComponent = container.mount.find(AddTodo)

    const addTodoInput = AddTodoComponent.find('[data-test="add-todo-input"]')
    const addTodoBtn = AddTodoComponent.find('[data-test="add-todo-btn"]')

    addTodoInput.simulate('change', { target: { value: addTodo } })

    expect(addTodoInput.instance().value).toEqual(addTodo)

    addTodoBtn.simulate('click')

    expect(container.store.getState().todos).toHaveLength(1)
    expect(container.store.getState().todos[0].id).toEqual(0)
  })

  it('does not add a empty todo item', () => {
    const todos = []
    const addTodo = ''

    const container = renderAppComponent(
      {
        todos,
      },
      { route: '/todos' }
    )

    const TodoPage = container.mount
    const newTodoBtn = TodoPage.find('[data-test="new-todo-btn"]')

    newTodoBtn.simulate('click')

    const addTodoInput = TodoPage.find('[data-test="add-todo-input"]')
    const addTodoBtn = TodoPage.find('[data-test="add-todo-btn"]')

    addTodoInput.simulate('change', { target: { value: addTodo } })

    expect(addTodoInput.instance().value).toEqual(addTodo)

    addTodoBtn.simulate('click')

    expect(container.store.getState().todos).toHaveLength(0)
  })

  it('can cancel input for new todo item', () => {
    const todos = []
    const addTodo = 'clean my room'

    const container = renderAppComponent(
      {
        todos,
      },
      { route: '/todos' }
    )

    const TodoPage = container.mount
    const newTodoBtn = TodoPage.find('[data-test="new-todo-btn"]')
    newTodoBtn.simulate('click')

    const cancelTodoBtn = TodoPage.find('[data-test="cancel-todo-btn"]')

    expect(TodoPage.find('[data-test="add-todo"]')).toHaveLength(1)

    cancelTodoBtn.simulate('click')

    expect(TodoPage.find('div[data-test="add-todo"]')).toHaveLength(0)

    expect(container.store.getState().todos).toHaveLength(0)
  })
})
