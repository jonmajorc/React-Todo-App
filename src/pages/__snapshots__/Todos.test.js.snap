// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos Page renders todo page 1`] = `
<div
  className="todos"
  data-test="todo-page"
>
  <header
    className="todos__header"
  >
    <h3>
      My Todos
    </h3>
    <small
      className="todos__header__date"
    >
      Fri May 25 2018
    </small>
    <button
      className="todos__new-todo-btn"
      data-test="new-todo-btn"
      onClick={[Function]}
    >
      + Add Todo
    </button>
  </header>
  <small
    className="todos__header__help"
  >
    ( Click a todo to complete it! )
  </small>
  <TodoList
    removeTodo={[Function]}
    todos={Array []}
    toggleTodo={[Function]}
  >
    <ul
      className="todo-list"
    />
  </TodoList>
</div>
`;
